app/src/main/java/org/kiwix/kiwixmobile/zim_manager/local_file_transfer/WifiDirectManager.java

```diff
@@ -0,0 +1,489 @@
+  private ArrayList<FileItem> filesForTransfer = new ArrayList<>();
+
+  private ArrayList<Uri> fileUriArrayList; // For sender device, stores uris of the files
+  private boolean isFileSender = false;    // Whether the device is the file sender or not
+
+  public WifiDirectManager(@NonNull LocalFileTransferActivity activity) {
+    this.activity = activity;
+  }
+
+  /* Initialisations for using the WiFi P2P API */
+  public void createWifiDirectManager(@NonNull SharedPreferenceUtil sharedPreferenceUtil,
+      @NonNull AlertDialogShower alertDialogShower, @Nullable ArrayList<FileItem> filesForTransfer,
+      @Nullable ArrayList<Uri> fileUriArrayList) {
+    this.sharedPreferenceUtil = sharedPreferenceUtil;
+    this.alertDialogShower = alertDialogShower;
+    this.fileUriArrayList = fileUriArrayList;
+    this.filesForTransfer = filesForTransfer;
+    this.isFileSender = (fileUriArrayList != null && fileUriArrayList.size() > 0);
+
+    if(isFileSender) {
+      this.totalFilesForTransfer = fileUriArrayList.size();
+    }
+
+    manager = (WifiP2pManager) activity.getSystemService(Context.WIFI_P2P_SERVICE);
+    channel = manager.initialize(activity, getMainLooper(), null);
+  }
+
+  public void registerWifiDirectBroadcastReceiver() {
+    receiver = new KiwixWifiP2pBroadcastReceiver(this);
+
+    // For specifying broadcasts (of the P2P API) that the module needs to respond to
+    IntentFilter intentFilter = new IntentFilter();
+    intentFilter.addAction(WifiP2pManager.WIFI_P2P_STATE_CHANGED_ACTION);
+    intentFilter.addAction(WifiP2pManager.WIFI_P2P_PEERS_CHANGED_ACTION);
+    intentFilter.addAction(WifiP2pManager.WIFI_P2P_CONNECTION_CHANGED_ACTION);
+    intentFilter.addAction(WifiP2pManager.WIFI_P2P_THIS_DEVICE_CHANGED_ACTION);
+
+    activity.registerReceiver(receiver, intentFilter);
+  }
+
+  public void unregisterWifiDirectBroadcastReceiver() {
+    activity.unregisterReceiver(receiver);
+  }
+
+  public void discoverPeerDevices() {
+    manager.discoverPeers(channel, new WifiP2pManager.ActionListener() {
+      @Override
+      public void onSuccess() {
+        displayToast(R.string.discovery_initiated, Toast.LENGTH_SHORT);
+      }
+
+      @Override
+      public void onFailure(int reason) {
+        String errorMessage = getErrorMessage(reason);
+        Log.d(TAG, activity.getString(R.string.discovery_failed) + ": " + errorMessage);
+        displayToast(R.string.discovery_failed, Toast.LENGTH_SHORT);
+      }
+    });
+  }
+
+  @Override
+  public void onWifiP2pStateChanged(int wifiP2pState) {
+    this.isWifiP2pEnabled = (wifiP2pState == WifiP2pManager.WIFI_P2P_STATE_ENABLED);
+
+    if(!isWifiP2pEnabled) {
+      displayToast(R.string.discovery_needs_wifi, Toast.LENGTH_SHORT);
+      ((Callbacks) activity).clearListOfAvailablePeers();
+    }
+
+    Log.d(TAG, "WiFi P2P state changed - " + isWifiP2pEnabled);
+  }
+
+  @Override
+  public void onPeersChanged() {
+    if (manager != null) {
```

Comment:
I wouldn't null check here, if this field is null and you call this method then the app should crash because something crazy has happened. The other possibility is that the class is being used wrong in devlopment and instead of a crash happening we just get a log statement

Reply:
I have a general query about development: For an app-flow which will never normally happen (like manager being null here), is it always okay to let the app crash in such a freak scenario?

Reply:
Absolutely, https://proandroiddev.com/handling-exceptions-the-easy-way-3aeade554613 see handy dandy graph here. If something impossible happened I damn well want to know about it and fix it.

Kotlin has some quite cool little functions for scenarios like [this](https://www.bignerdranch.com/blog/write-better-code-using-kotlins-require-check-and-assert/)

Reply:
I'll checkout the links, thanks
[Removed](https://github.com/kiwix/kiwix-android/pull/1276/commits/d78f07be6abfc481c8c46b870062346de3d0118c) the null checks